text/0626-invoke-helper.md

```diff
@@ -0,0 +1,475 @@
+      this.result = await response.json();
+    } catch {
+      this.isError = true;
+    } finally {
+      this.isLoading = false;
+    }
+  }
+
+  refresh() {
+    this.run();
+  }
+}
+
+class RemoteData extends Helper {
+  compute([url]) {
+    return new FetchTask(url);
+  }
+}
+
+
+class Poll extends Helper {
+  intervalId = null;
+
+  compute([callback, pollPeriod]) {
+    cancelInterval(this.intervalId);
+
+    this.intervalId = setInterval(callback, pollPeriod);
+  }
+
+  willDestroy() {
+    cancelInterval(this.intervalId);
+  }
+}
+
+export default class Search extends Component {
+  data = invokeHelper(this, RemoteData, () => ({
+    positional: [`www.example.com/search?query=${this.text}`]
+  }));
+
+  constructor() {
+    super(...arguments);
+
+    invokeHelper(this, Poll, () => ({
+      positional: [() => this.data.value.refresh(), this.pollPeriod]
+    }));
+  }
+}
+```
+
+Note how the concerns are completely separated in this version of the component.
+The polling logic is self contained, and separated from the data fetching logic.
+Both sets of logic are able to contain their lifecycles, updating based on
+changes to tracked state, and tearing down when the program destroys them. In
+the future, convenience APIs can be added to make invoking them easier to read:
+
+```js
+export default class Search extends Component {
+  @use data = remoteData(() => `www.example.com/search?query=${this.text}`);
+
+  constructor() {
+    super(...arguments);
+
+    use(this, poll(() => [() => this.data.refresh(), this.pollPeriod]));
+  }
+}
+```
+
+## Dependencies
+
+This RFC depends on the [Helper Manager RFC](https://github.com/emberjs/rfcs/pull/625).
+
+## Detailed design
+
+This RFC proposes adding the `invokeHelper` function, imported from
+`@ember/helper`. The function will have the following interface (using
+TypeScript types for brevity and clarity):
+
+```ts
+interface TemplateArgs {
+  positional?: unknown[],
+  named?: Record<string, unknown>
+}
+
+type HelperDefinition = object;
+
+interface Helper {
+  value: unknown;
+}
+
+function invokeHelper(
+  context: object,
+  definition: HelperDefinition,
+  argsGetter?: (context: object) => TemplateArgs
+): Helper;
+```
+
+Let's step through the arguments to the function one by one:
+
+#### `context`
+
+This is the parent context for the helper definition. The helper will be
+associated as a destroyable to this parent context, using the destroyables API,
+so that its lifecycle is tied to the parent. The only requirement of the parent
+is that is an object of some kind that can be destroyed.
+
+This allows helper's lifecycles to be entangled correctly with the parent, and
+encourages users to ensure they've properly handled the lifecycle of their
+helper.
+
+#### `definition`
+
+This is the helper definition. It can be any object, with the only requirement
+being that a helper manager has been associated with it via the
+`setHelperManager` API.
+
+#### `argsGetter`
```

Comment:
Can we call this `computeFunction` or something? I think of "getter" generally in the ES5 `get someFunction() {}` sense...